### PR TITLE
`Communication`: Add answer posts content to course wide search

### PR DESCRIPTION
**Pull request by cremertim**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
Currently, the search only filters the content of the original posts

### Description
I added answer posts, where the original post does not match the content but an answer post matches, to the search result

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course

1. Log in to Artemis
2. Navigate into the course
3. If not already existent, create a message in a channel with an answer post and content x
4. Search for content x of the answer post
5. Verify search result includes post, where the answer post matches the content

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
- [x] Test 3
- [x] Test 4
- [x] Test 5

## Summary by CodeRabbit

- **New Features**
	- Enhanced message filtering to include both posts and answer posts based on search criteria.
	- Improved retrieval of user-specific posts and associated answer posts.

- **Tests**
	- Introduced new test methods to ensure robust coverage of search functionalities in messaging, validating retrieval based on specific search criteria for both posts and answers.

### CODE REVIEW

**Review by HawKhiem — COMMENTED** [review]
Tested on TS3. I presume you meant messages in communication by "posts". Please correct me if I was testing the wrong thing. Everything works as described. One small thing though: After done searching for something then click on the Refresh button, I expect the origin view of the channel would show up with all the messages. But instead still my search result show up, it would only revert back to the original view after I click on another channel then go back, or by emptying the search bar and click on the search symbol (which happened to work in this case because I don't have any other public posts on other channel)


https://github.com/user-attachments/assets/c8039eac-f7fc-4e86-821f-003a641f1416

File: src/main/java/de/tum/cit/aet/artemis/communication/repository/MessageSpecs.java [hunk 1]
```diff
@@ -114,7 +116,9 @@ public static Specification<Post> getOwnSpecification(boolean filterToOwn, Long
                 return null;
             }
             else {
-                return criteriaBuilder.equal(root.get(Post_.AUTHOR).get(User_.ID), userId);
+                Join<Post, AnswerPost> answersJoin = root.join(Post_.ANSWERS, JoinType.LEFT);
+                Predicate searchInAnswerContent = criteriaBuilder.equal(answersJoin.get(AnswerPost_.AUTHOR).get(User_.ID), userId);
+                return criteriaBuilder.or(criteriaBuilder.equal(root.get(Post_.AUTHOR).get(User_.ID), userId), searchInAnswerContent);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/repository/MessageSpecs.java:121]:
```suggestion
                someMeaningfulVariable = criteriaBuilder.equal(root.get(Post_.AUTHOR).get(User_.ID), userId);
                return criteriaBuilder.or(someMeaningfulVariable, searchInAnswerContent);
```
Can you extract the first criteria as well so others can understand which two criteria are used?

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good apart from one difficult to understand part

File: src/main/java/de/tum/cit/aet/artemis/communication/repository/MessageSpecs.java (hunk 1 above)
Comment (cremertim) [reply src/main/java/de/tum/cit/aet/artemis/communication/repository/MessageSpecs.java:121]: Done

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by sachmii — APPROVED** [review]
Tested on TS5, worked fine.

**Review by SindiBuklaji — COMMENTED** [review]
Tested on TS5. Everything works great. 

One remark though, when searching for a message and that message is found within a thread, it does find the correct thread, but it doesn't show you the actual message.

image

**Review by SindiBuklaji — APPROVED** [review]
Reapprove

**Review by Cathy0123456789 — COMMENTED** [review]
Tested on TS3. Works as described.
Only searching for % doesn't give the correct result and shows some messages twice
image

**Review by zagemello — COMMENTED** [review]
The _ underline sign is identified as a universal sign, a "joker" basically, meaning that searching a word and putting it in for any letter will still return the right search. Is consistent across all artemis searches. 
image

**Review by HanyangXu0508 — APPROVED** [review]
I tested on TS2 and all works as expected.

**Review by asliayk — APPROVED** [review]
tested locally and works as expected

**Review by kevinfischer4 — APPROVED** [review]
Tested on TS1, works as expected. Cool feature!
Screenshot 2024-11-03 at 19 42 29

**Review by krusche — APPROVED** [review]
Code looks good to me. We can implement additional search improvements in a follow-up

**Pull request merged**